addons/nuxeo-retention/nuxeo-retention-core/src/main/java/org/nuxeo/retention/adapters/RetentionRule.java

```diff
@@ -0,0 +1,246 @@
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * @since 11.1
+ */
+public class RetentionRule {
+
+    public enum ApplicationPolicy {
+        AUTO, MANUAL
+    }
+
+    public enum StartingPointPolicy {
+        IMMEDIATE, AFTER_DELAY, EVENT_BASED, METADATA_BASED
+    }
+
+    private static final Logger log = LogManager.getLogger(Record.class);
+
+    protected DocumentModel document;
+
+    public RetentionRule(DocumentModel doc) {
+        this.document = doc;
+    }
+
+    public void disable() {
+        document.setPropertyValue(RetentionConstants.ENABLED_PROP, false);
+    }
+
+    public void enable() {
+        document.setPropertyValue(RetentionConstants.ENABLED_PROP, true);
+    }
+
+    public String getApplicationPolicy() {
+        return (String) document.getPropertyValue(RetentionConstants.APPLICATION_POLICY_PROP);
+    }
+
+    public List<String> getBeginActions() {
+        Serializable propertyValue = document.getPropertyValue(RetentionConstants.BEGIN_ACTIONS_PROP);
+        if (propertyValue == null) {
+            return Collections.emptyList();
+        }
+        return Arrays.asList((String[]) propertyValue);
+    }
+
+    public List<String> getDocTypes() {
+        @SuppressWarnings("unchecked")
+        List<String> propertyValue = (List<String>) document.getPropertyValue(RetentionConstants.DOC_TYPES_PROP);
+        return propertyValue;
+    }
+
+    public DocumentModel getDocument() {
+        return document;
+    }
+
+    public Long getDurationDays() {
+        return (Long) document.getPropertyValue(RetentionConstants.DURATION_DAYS_PROP);
+    }
+
+    public Long getDurationMillis() {
+        return (Long) document.getPropertyValue(RetentionConstants.DURATION_MILLIS_PROP);
+    }
+
+    public Long getDurationMonths() {
+        return (Long) document.getPropertyValue(RetentionConstants.DURATION_MONTHS_PROP);
+    }
+
+    public Long getDurationYears() {
+        return (Long) document.getPropertyValue(RetentionConstants.DURATION_YEARS_PROP);
+    }
+
+    public List<String> getEndActions() {
+        Serializable propertyValue = document.getPropertyValue(RetentionConstants.END_ACTIONS_PROP);
+        if (propertyValue == null) {
+            return Collections.emptyList();
+        }
+        return Arrays.asList((String[]) propertyValue);
+    }
+
+    public String getExpression() {
+        return (String) document.getPropertyValue(RetentionConstants.EXPRESSION_PROP);
+    }
+
+    public String getMetadataXpath() {
+        return (String) document.getPropertyValue(RetentionConstants.METADATA_XPATH_PROP);
+    }
+
+    public Calendar getRetainUntilDateFrom(Calendar calendar) {
+        LocalDateTime datetime = LocalDateTime.ofInstant(calendar.getTime().toInstant(), ZoneId.systemDefault());
+        return getRetainUntilDateFrom(datetime);
+    }
+
+    protected Calendar getRetainUntilDateFrom(LocalDateTime datetime) {
+        LocalDateTime localDateTime = datetime.plusYears(getDurationYears())
```

Comment:
can we make a new variable from datetime parameter to enforce immutability and defensive copy = > `LocalDateTime localDateTime = LocalDateTime.from(datetime);`

Reply:
They're already immutable, `plusYears` and friends already return a copy.